ygot/struct_validation_map.go

```diff
@@ -319,3 +321,189 @@ func MergeJSON(a, b map[string]interface{}) (map[string]interface{}, error) {
 
 	return o, nil
 }
+
+// MergeStructs takes two input ValidatedGoStructs and merges their contents,
+// returning a new ValidatedGoStruct. If the input structs a and b are of
+// different types, an error is returned.
+//
+// In the case that the structs contain a slice, or a map that is already
+// populated in both structs, an error is returned. Merging two lists with
+// identical members will be added in future iterations of this code.
+//
+// TODO(robjs): Fix the unimplemented test cases where two structs of
+// the same type have slices or maps that are already populated.
+// See https://github.com/openconfig/ygot/issues/74.
+func MergeStructs(a, b ValidatedGoStruct) (ValidatedGoStruct, error) {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return nil, fmt.Errorf("cannot merge structs that are not of matching types, %T != %T", a, b)
+	}
+
+	n := reflect.New(reflect.TypeOf(a).Elem())
+	if err := copyStruct(n.Elem(), reflect.ValueOf(a).Elem()); err != nil {
+		return nil, fmt.Errorf("error merging a to new struct: %v", err)
+	}
+
+	if err := copyStruct(n.Elem(), reflect.ValueOf(b).Elem()); err != nil {
+		return nil, fmt.Errorf("error merging b to new struct: %v", err)
+	}
+
+	return n.Interface().(ValidatedGoStruct), nil
+
+}
+
+// copyStruct copies the fields of srcVal into the dstVal struct in-place.
+func copyStruct(dstVal, srcVal reflect.Value) error {
+	if srcVal.Type() != dstVal.Type() {
+		return fmt.Errorf("cannot copy %s to %s", srcVal.Type().Name(), dstVal.Type().Name())
+	}
+
+	if !util.IsValueStruct(dstVal) || !util.IsValueStruct(srcVal) {
+		return fmt.Errorf("cannot handle non-struct types, src: %v, dst: %v", srcVal.Type().Kind(), dstVal.Type().Kind())
+	}
+
+	for i := 0; i < srcVal.NumField(); i++ {
+		srcField := srcVal.Field(i)
+		dstField := dstVal.Field(i)
+
+		switch srcField.Kind() {
+		case reflect.Ptr:
+			if err := copyPtrField(dstField, srcField); err != nil {
+				return err
+			}
+		case reflect.Interface:
+			if err := copyInterfaceField(dstField, srcField); err != nil {
+				return err
+			}
+		case reflect.Map:
+			if err := copyMapField(dstField, srcField); err != nil {
+				return err
+			}
+		case reflect.Slice:
+			if err := copySliceField(dstField, srcField); err != nil {
+				return err
+			}
+		default:
+			dstField.Set(srcField)
+		}
+	}
+	return nil
+}
+
+// copyPtrField copies srcField to dstField. srcField and dstField must be
+// reflect.Value structs which represent pointers.
+func copyPtrField(dstField, srcField reflect.Value) error {
+	if srcField.IsNil() || !srcField.IsValid() {
+		return nil
+	}
+
+	switch srcField.Elem().Kind() {
+	case reflect.Struct:
```

Comment:
if util.IsValueStructPtr(srcField) { // avoid potential panic on .Elem()
 blah
 return nil
} 
// default case

Reply:
I'd erred on the side of keeping these internal functions doing less validation. Since this is internal and we'd already checked for `reflect.Ptr` before calling this function (it's really there to divide the logic) then this seems to be a little more checking than might be required.

Have updated the function - but can't just do this is we want to avoid the panic, since we need to call `.Elem()` in both cases to get the thing pointed to, so I've added two cases.